text/0028-scaling-codeowners.md

```diff
@@ -0,0 +1,63 @@
+
+# Summary
+
+Sentry supports GitHub and Gitlab Code Owners files for auto-assigning Issues. However, Sentry only supports 100K characters as the file size limit for Code Owners. GitHub supports a 3Mb file size for Code Owners.
+
+This is increasingly becoming a pain point for Enterprise customers who want to use their Code Owners files as is.
+
+# Motivation
+
+There are two reasons behind the importance of this change
+
+### Pain Point for Enterprise Customers
+
+Enterprise Customers want to use their existing Code Owners files and do not want to reduce their file sizes using wild cards just because Sentry cannot handle the same size as Github. In the past, we already made an exception for select orgs and support a 200K char file size for them.
+
+### Code Owners Increases Auto Assignment
+
+As we make efforts to increase the signal-to-noise ratio in our notifications and Issue stream - Being able to answer “Who is this Issue important for?” is a necessary opinion that Sentry needs to build. Code Owners enables auto assignment of Issues which is one of the clearest signals of issue ownership we have today. We want more auto assignment of Issues in Sentry and thus we need more Orgs to use Code Owners
+
+# Goals:
+
+1. Users can upload a Code Owners File Size that's up to 3Mb in size
+2. The processing time to calculate assignments should not increase significantly. 
+
+# How it Works (simplified)
+
+We have Code Mappings that map the transformation of the source code to the production build. Code Mappings help us reverse-engineer the file paths from an error’s stacktrace to source code file paths. 
+
+A customer can apply a Code Mapping to a CODEOWNERS file to get the file paths that should match against the stacktrace file paths. The transformed output is stored in `sentry_projectcodeowners` and is what Sentry will use to determine assignments.
+
+When Sentry ingests a new event, we go through each line in our transformed CODEOWNERS and see if each line has a match across any of the frames in the stacktrace. We only use the last matching line for the assignment.
+
+# Current System Constraints
+
+- We store the original and transformed CODEOWNERS file in a `sentry_projectcodeowners` row. However to prevent overloading Postgres from queries on every event, we cache the row in Memcache. Memcache can only support 1mb. Rows larger than 1mb will hit the database directly. Since we store the original and transformed CODEOWNERS in the same row, files larger than 500k chars will exceed this limitation.
+- On each event, the assignment calculation is currently O((m*l + o)*n) time complexity, where *m* is the # of lines in the CODEOWNERS file, *l* is the number of code mappings mapped to the project, *o* is the number of lines in the Ownership Rules and n is the number of lines in the stacktrace. Essentially quadratic.
+- We recalculate the assignment for each event, even if the Ownership Rules and CODEOWNERS file has not been updated.
```

Comment:
It's hard to avoid this to some extent, because we could have different events in a single issue.